internal/bufpushaction/bufpushaction.go

```diff
@@ -0,0 +1,265 @@
+		input:            input,
+		track:            track,
+		stdout:           container.Stdout(),
+		stderr:           container.Stderr(),
+		currentGitCommit: currentGitCommit,
+		moduleName:       moduleName,
+		githubClient:     githubClient,
+		bufRunner: &bufRunner{
+			bufToken: container.Env(bufTokenKey),
+			path:     container.Env("PATH"),
+		},
+	}.push(ctx)
+}
+
+// interceptErrorForGithubAction intercepts errors and wraps them in formatting required for an error to be shown in
+// the workflow results.
+func interceptErrorForGithubAction(
+	next func(context.Context, appflag.Container) error,
+) func(context.Context, appflag.Container) error {
+	return func(ctx context.Context, container appflag.Container) error {
+		err := next(ctx, container)
+		if err == nil {
+			return nil
+		}
+		return fmt.Errorf("::error::%v", err)
+	}
+}
+
+func getNameFromConfigFile(ctx context.Context, bucket storage.ReadBucket) (_ string, retErr error) {
+	file, err := bucket.Get(ctx, "buf.yaml")
+	if err != nil {
+		return "", err
+	}
+	defer func() {
+		retErr = multierr.Append(retErr, file.Close())
+	}()
+	name := struct {
+		Name string `yaml:"name"`
+	}{}
+	err = yaml.NewDecoder(file).Decode(&name)
+	if err != nil {
+		return "", err
+	}
+	return name.Name, nil
+}
+
+type pusher struct {
+	input            string
+	track            string
+	stdout           io.Writer
+	stderr           io.Writer
+	currentGitCommit string
+	moduleName       string
+	githubClient     github.Client
+	bufRunner        commandRunner
+}
+
+var errNoTrackSupport = errors.New("The installed version of buf does not support setting the track. Please use buf v1.0.0 or newer.")
+
+func (p pusher) push(ctx context.Context) error {
+	// versions of buf prior to --track support emit this message when running `buf push --track foo --help`
+	const unsupportedTrackHelpFragment = "unknown flag: --track"
+
+	// make sure --track is supported
+	if p.track != "main" {
+		_, stderr, err := p.bufRunner.Run(ctx, "push", "--track", p.track, "--help")
+		if err != nil {
+			if strings.Contains(stderr, unsupportedTrackHelpFragment) {
+				return errNoTrackSupport
+			}
+			return err
+		}
+	}
+	tags, err := p.getTags(ctx)
+	if err != nil {
+		return err
+	}
+	for _, tag := range tags {
+		if len(tag) != 40 {
+			continue
+		}
+		if _, err := hex.DecodeString(tag); err != nil {
+			continue
+		}
+		status, err := p.githubClient.CompareCommits(ctx, tag, p.currentGitCommit)
+		if err != nil {
+			if github.IsNotFoundError(err) {
+				continue
+			}
+			return err
+		}
+		switch status {
+		case github.CompareCommitsStatusIdentical:
+			p.notice(fmt.Sprintf("Skipping because the current git commit is already the head of track %s", p.track))
+			return nil
+		case github.CompareCommitsStatusBehind:
+			p.notice(fmt.Sprintf("Skipping because the current git commit is behind the head of track %s", p.track))
+			return nil
+		case github.CompareCommitsStatusDiverged, github.CompareCommitsStatusAhead:
```

Comment:
I went back and forth on how to handle diverged.

There are two likely scenarios where this will happen:
1. Somebody force-pushed to the git branch.
2. The git branch was deleted and a new one was created with the same name. This will normally not happen because we will recommend that a workflow be configured to delete the BSR track when a GitHub branch is deleted.

In both of those situations, pushing the module seems like the right thing to do.

Reply:
I think I agree with you that pushing in those scenarios makes sense. However, I think we should maybe print a `notice` message saying we detected they diverged but that the default behavior is to still push. That way if there's a scenario we haven't considered this decision would be visible to the user, rather than silently being treated the same as ahead.